jobs: write to new jobs tables

dt opened the pull request:
See commits.

dt committed:
jobs: add progress and status table accessors

Release note: none.
Epic: none.

dt committed:
jobs: extend GC and delete funcs to cleanup new tables

Release note: none.
Epic: none.

dt committed:
jobs: write to new tables during updates

Release note: none.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137218)
<!-- Reviewable:end -->

msbutler commented on pkg/jobs/update.go:
```diff
@@ -263,6 +266,45 @@ WHERE id = $1
+	}
+	if v.AtLeast(clusterversion.V25_1_AddJobsTables.Version()) {
+		if ju.md.Status != "" && ju.md.Status != status {
+			if err := j.StatusStorage().Record(ctx, u.txn, "state", string(ju.md.Status)); err != nil {
```
i'd think that if the job changes from running to paused, we'd want to update the job_status, even if progress was nil?

msbutler commented on pkg/jobs/update.go:
```diff
@@ -263,6 +266,45 @@ WHERE id = $1
+					return err
+				}
+			} else if ju.md.Status != "" && ju.md.Status != status {
+				if err := j.StatusStorage().Clear(ctx, u.txn); err != nil {
```
what's the motivation for this branch? I don't follow why if `progress != nil` and the job changed from state without changing running status, we need to clear the status.

dt replied:
Anytime the job changes state, we want to clear the status from the old state.

But if we're going to set it to something new in the same txn we don't need to clear it twice. We only know if we're setting it to something new if progress -- where the new thing could be -- is non-nil.

I'll rework to move the Clear to the above case.

dt committed:
upgrades: use extra versions for upgrade tests

Previously some of these tests were replacing existing upgrades between min and latest
with their overridden impl, which resulted in the actual upgrade not running even though
the cluster and schema versions would be updated to indicate it had. This causes problems
for any systems in the cluster that then switch on that version and expect its effects
when called by the systems run in these tests.

Instead these tests now create additional versions beyond Latest. This is its own can
of worms and out-of-bounds errors but usage is currently confined to these tests and hopefully
can stay that way.

Release note: none.
Epic: none.

dt commented on pkg/jobs/job_info_storage.go:
```diff
@@ -8,18 +8,304 @@ package jobs
+		return err
+	}
+
+	if _, err := txn.ExecEx(
```
up to 1 (probably zero) yeah. the prune should be no worse than the delete we do of the single row.

dt commented on pkg/jobs/job_info_storage.go:
```diff
@@ -8,18 +8,304 @@ package jobs
+	Written       time.Time
+}
+
+func (i StatusStorage) AllMessages(
```
What user? I don't think we want anyone "reading the <system> table" -- we want them using our accessors. But the API is mostly so I can write a test.

dt commented on pkg/jobs/jobs_test.go:
```diff
@@ -19,6 +19,7 @@ import (
 	"time"
 
 	apd "github.com/cockroachdb/apd/v3"
+	_ "github.com/cockroachdb/cockroach/pkg/backup" // we use backup jobs as placeholders.
```
it was already linked in; I just moved the import to the correct file.

dt commented on pkg/jobs/job_info_storage.go:
```diff
@@ -8,18 +8,304 @@ package jobs
+}
+
+// Record writes a human readable message of the specified kind to the message
+// log for this job, and prunes retained messages of the same kind based on the
```
Eh, I don't love centralized lists that need to be updated for everyone anyone anyone wants to add a thing. Why do we care?

msbutler replied:
i don't feel strongly about this, so i'm fine letting users have their own kinds

dt commented on pkg/jobs/registry.go:
```diff
@@ -1177,6 +1178,12 @@ SELECT distinct (id), latestpayload.value AS payload, status
 FROM jobpage AS j
 INNER JOIN latestpayload ON j.id = latestpayload.job_id`
 
+// jobMetadataTables all all of the tables that have rows for a given job that
```
they're the job *meta* data tables, i.e. data _about_ the jobs, not the jobs themselves?

msbutler commented on pkg/jobs/job_info_storage.go:
```diff
@@ -8,18 +8,317 @@ package jobs
+	ctx, sp := tracing.ChildSpan(ctx, "get-job-progress")
+	defer sp.Finish()
+
+	// The sort and limit should not be required since Write deletes all prior
```
stray comment.

msbutler commented on pkg/jobs/update.go:
```diff
@@ -263,6 +266,48 @@ WHERE id = $1
+				ts = *hwm
+			}
+
+			if err := j.ProgressStorage().Set(ctx, u.txn, float64(progress.GetFractionCompleted()), ts); err != nil {
```
should we check if progress has advanced before writing to `ProgressStorage`?

msbutler commented on pkg/jobs/job_info_storage.go:
```diff
@@ -8,18 +8,317 @@ package jobs
+		return err
+	}
+
+	if err := MessageStorage(i).Record(ctx, txn, "status", status); err != nil {
```
we could disallow users from using "status" , but i don't care strongly.

msbutler commented on pkg/jobs/job_info_storage.go:
```diff
@@ -8,18 +8,317 @@ package jobs
+// that are for *human* consumption. These are distinct from info rows, which
+// are byte strings used by the code of the job itself to persist state that it
+// will consume, and distinct from the message log, as a job can only have one
+// "status" at a time, but messages are histoical log.
```
it may be worth clarifying here that StatusStorage will only store running status messages, and not the job state (e.g. running vs paused).

msbutler reviewed (approved):
looking forward to more tests!

dt commented:
> more tests
Anything in particular? Since we're just moving stuff around, rather than adding new functionality (except the message log but that has no callers so far), I think the main "test" right now is just that the current tests still pass, i.e. that we haven't regressed the externally visible functionality, and on that front we're good -- this is green.

msbutler commented:
> Anything in particular? 
I'm happy to merge this now, but at some point we should test the new pruning logic, job gc on the new tables.

I assume we'll test new update.go logic, once we update SHOW JOB(s).

dt commented:
>  once we update SHOW JOB(s).

Exactly; I think the tests still passing when we update the read path is the true proof we got this stuff right here. Other than the bounded retention of the history/message rows, everything here _should_ be exercised by existing tests that we just don't want to break.

TFTR!

bors r+
